### PR TITLE
Test modifications for FIPS 140 mode

### PR DESCRIPTION
**Pull request by jkakavas**
- Enable SunJGSS provider for Kerberos tests
- Handle the fact that in the decrypt method in KeyStoreWrapper might
not throw immediately when the GCM cipher is from BouncyCastle FIPS
and we end up with a DataInputStream that has reached it's end.
- Disable tests, jarHell, testingConventions for ingest attachment
plugin. We don't support this plugin (and document this) in FIPS
mode.
- Don't attempt to install ingest-attachment in smoke-test-plugins

### CODE REVIEW

**Comment by jkakavas** [comment]
Just completed a local CI run in FIPS mode with the changes included in this PR, so this will hopefully bring our fips matrix job to green for master.

**Review by mark-vieira — CHANGES_REQUESTED** [review]
Looks like we have some packaging tests failing here. We'll need to address that.

**Comment by jkakavas** [comment]
@mark-vieira the test is fixed

File: qa/os/src/test/java/org/elasticsearch/packaging/test/PackagingTestCase.java [hunk 1]
```diff
@@ -310,23 +316,28 @@ public void startElasticsearch() throws Exception {
...
+        Matcher<String>[] stringMatchers = expectedMessages.stream().map(CoreMatchers::containsString).toArray(Matcher[]::new);
         if (Files.exists(installation.logs.resolve("elasticsearch.log"))) {
 
             // If log file exists, then we have bootstrapped our logging and the
             // error should be in the logs
             assertTrue("log file exists", Files.exists(installation.logs.resolve("elasticsearch.log")));
             String logfile = FileUtils.slurp(installation.logs.resolve("elasticsearch.log"));
-            assertThat(logfile, containsString(expectedMessage));
+
+            assertThat(logfile, anyOf(Arrays.asList(stringMatchers)));
```
Comment (pugnascotia) [inline qa/os/src/test/java/org/elasticsearch/packaging/test/PackagingTestCase.java]: Should be the following instead? The other uses of `stringMatchers` in this method aren't wrapped in `Arrays.asList`.

```suggestion
            assertThat(logfile, anyOf(stringMatchers));
```

**Review by pugnascotia — APPROVED** [review]
LGTM. What causes the keystore error to vary between "incorrect" and "tampered with"?

File: qa/os/src/test/java/org/elasticsearch/packaging/test/PackagingTestCase.java (hunk 1 above)
Comment (jkakavas) [reply qa/os/src/test/java/org/elasticsearch/packaging/test/PackagingTestCase.java]: Both work as anyOf has multiple signatures, one that takes a vararg ( so we can pass an array of Matcher ) and it also has one that takes an Iterable ( so we can pass a List  ). That said, the Arrays.asList is a leftover from a previous attempt and we gain nothing from converting the Array to a List, so I'll change for consistency

**Comment by jkakavas** [comment]
> What causes the keystore error to vary between "incorrect" and "tampered with"?

See 
https://github.com/elastic/elasticsearch/blob/84dd9dc9c03d85dd08fdaa67ce78c71511e7a01d/server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java#L361-L367

`CipherInputStream` and `Cipher` implementations come from the Security Provider in use. When in FIPS mode that Provider is `BCFIPS` ( BouncyCastle in FIPS mode ) while when not in FIPS mode, that Provider is SunJCE. 

When the password is incorrect, thus the produced decryption key is incorrect the two implementations behave differently: 

1. SunJCE's CipherInputStream throws an `javax.crypto.AEADBadTagException` on `int numEntries = input.readInt();` which we catch and translate to "wrong password"

2. BCFIPS CipherInputStream reads bytes into an integer that doesn't make sense ( either negative or positive ) . 
    a. When the integer is positive ( i.e. `233254` ) we proceed to try and read all these "entries" from the keystore, but the next attempt to read from the `DataInputStream` 
        
     ```java
     String setting = input.readUTF();
     ```  
      throws an an `javax.crypto.AEADBadTagException` which we catch and traslate to "wrong password"

    b. When the integer is negative, we move to 
    ```java
    if (input.read() != -1) {
        hrow new SecurityException("Keystore has been corrupted or tampered with");
    }
    ```
    and since input.read() is not -1, we throw the "Keystore has been corrupted or tampered with"


I haven't had the time yet to look into detail into the CipherInputStream implementations and see _exactly_ what goes on and if this problematic or not and whether we should raise an issue with BouncyCastle folks, but because I don't see me getting to it in the next few days, a test fix that would allow us to run our CI in FIPS mode felt like an appropriate middle ground

**Comment by pugnascotia** [comment]
Thanks for the explanation 👍

**Review by mark-vieira — APPROVED** [review]
LGTM

**Pull request merged**